### PR TITLE
SONARJS-500 Remove EcmaScriptLexer from syntax highlighting

### PR DESCRIPTION
**Pull request by vilchik-elena**
Syntax highlighting was reimplemented using BaseTreeVisitor
Highlighting of numbers was added

### CODE REVIEW

File: javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java [hunk 1]
```diff
@@ -20,16 +20,17 @@
...
  */
 @Beta
-public interface SyntaxTrivia extends Tree {
+public interface SyntaxTrivia extends SyntaxToken {
 
   String comment();
 
   int startLine();
 
+  int column();
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java]: MAJOR Add the "@Override" annotation above this method signature [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

File: javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    this.offsets = new SourceFileOffsets(file, fileSystem.encoding());
+  }
+
+  private void highlight(int startOffset, int endOffset, String code) {
+    if (endOffset > startOffset) {
+      highlighting.highlight(startOffset, endOffset, code);
+    }
+  }
+
+  private boolean isKeyword(String text) {
```
Comment (SonarTech) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: MAJOR Make "isKeyword" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 3]
```diff
@@ -213,16 +213,16 @@ private void scanFile(InputFile inputFile, List<JavaScriptCheck> visitors, Issua
...
-      LOG.warn("Could not get " + Highlightable.class.getCanonicalName() + " for " + inputFile.file());
-    }
-  }
+//  private void highlight(InputFile inputFile, JavaScriptHighlighter highlighter) {
+//    Highlightable perspective = perspective(Highlightable.class, inputFile);
+//
+//    if (perspective != null) {
+//      highlighter.highlight(perspective, inputFile.file());
+//
+//    } else {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 4]
```diff
@@ -213,16 +213,16 @@ private void scanFile(InputFile inputFile, List<JavaScriptCheck> visitors, Issua
...
-    Highlightable perspective = resourcePerspectives.as(Highlightable.class, inputFile);
-
-    if (perspective != null) {
-      highlighter.highlight(perspective, inputFile.file());
-
-    } else {
-      LOG.warn("Could not get " + Highlightable.class.getCanonicalName() + " for " + inputFile.file());
-    }
-  }
+//  private void highlight(InputFile inputFile, JavaScriptHighlighter highlighter) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 5]
```diff
@@ -213,16 +213,16 @@ private void scanFile(InputFile inputFile, List<JavaScriptCheck> visitors, Issua
...
+//    Highlightable perspective = perspective(Highlightable.class, inputFile);
+//
+//    if (perspective != null) {
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 6]
```diff
@@ -139,7 +139,7 @@ public void analyse(Project project, SensorContext context) {
 
     for (InputFile inputFile : fileSystem.inputFiles(mainFilePredicate)) {
       analyse(inputFile, treeVisitors);
-      highlight(inputFile, highlighter);
+//      highlight(inputFile, highlighter);
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java [hunk 7]
```diff
@@ -122,9 +121,10 @@ public void analyse(Project project, SensorContext context) {
     List<JavaScriptCheck> treeVisitors = Lists.newArrayList();
 
     EcmaScriptConfiguration configuration = new EcmaScriptConfiguration(fileSystem.encoding());
-    JavaScriptHighlighter highlighter = new JavaScriptHighlighter(configuration);
+//    JavaScriptHighlighter highlighter = new JavaScriptHighlighter(configuration);
```
Comment (SonarTech) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/JavaScriptSquidSensor.java]: MAJOR This block of commented-out lines of code should be removed. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues:
* MAJOR 7 major

Watch the comments in this conversation to review them.

File: javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        }
+        highlight(offsets.startOffset(trivia), offsets.endOffset(trivia), code);
+      }
+    }
+  }
+
+  void initHighlighting(File file) {
+    InputFile inputFile = fileSystem.inputFile(fileSystem.predicates().is(file));
+    if (inputFile == null) {
+      throw new IllegalArgumentException("Cannot get " + Highlightable.class.getCanonicalName() + " for a null file");
```
Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: @vilchik-elena (cc @pynicolas ) I am wondering: are we sure that we want to fail the analysis in this case and at line 121? What about displaying an error log ?

File: javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java [hunk 9]
```diff
@@ -20,16 +20,14 @@
...
 /**
  * Represents a Trivia in the SyntaxTree.
  */
 @Beta
-public interface SyntaxTrivia extends Tree {
+public interface SyntaxTrivia extends SyntaxToken {
 
   String comment();
 
   int startLine();
```
Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java]: @vilchik-elena maybe this one is not needed anymore as we now SyntaxTrivia is extending SyntaxToken ? There is already "line" method ?

File: javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java (hunk 8 above)
Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: BTW @vilchik-elena previous behaviour was to log a warn, see [JavaScriptSquidSensor.java#L223](https://github.com/SonarCommunity/sonar-javascript/pull/75/files#diff-800b35cf5d8e3b14f2c3ae99d1634dc4L223)

**Comment by lindamartin** [comment]
@vilchik-elena see my comments, otherwise LGTM!

File: javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java (hunk 9 above)
Comment (vilchik-elena) [inline javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java]: @lindamartin i was lazy to replace usages:) gone in bd27285

File: javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java (hunk 8 above)
Comment (vilchik-elena) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: @lindamartin agree

Comment (vilchik-elena) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: @lindamartin Fixed in efd751d. Could you have a look?

Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/javascript/highlighter/HighlighterVisitor.java]: @vilchik-elena perfect!

File: javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java (hunk 9 above)
Comment (lindamartin) [inline javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/lexical/SyntaxTrivia.java]: @vilchik-elena Haha perfect!

**Pull request merged**